In [1]:
import pandas as pd
import numpy as np

In [38]:
data = pd.read_csv('Data201819SerieA.csv')
data.head()

Div        Date HomeTeam  AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  ...  \
0  I1  18/08/2018   Chievo  Juventus     2     3   A     1     1   D  ...   
1  I1  18/08/2018    Lazio    Napoli     1     2   A     1     1   D  ...   
2  I1  19/08/2018  Bologna      Spal     0     1   A     0     0   D  ...   
3  I1  19/08/2018   Empoli  Cagliari     2     0   H     1     0   H  ...   
4  I1  19/08/2018    Parma   Udinese     2     2   D     1     0   H  ...   

   BbAv<2.5  BbAH  BbAHh  BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA   PSCH  PSCD  \
0      2.13    19   2.00     1.68     1.64     2.38     2.29  18.84  6.42   
1      2.17    20   0.00     2.12     2.07     1.83     1.79   2.78  3.57   
2      1.58    19  -0.25     1.97     1.92     1.99     1.94   2.31  3.18   
3      1.71    19  -0.25     1.98     1.91     1.98     1.94   2.54  3.42   
4      1.65    20   0.00     1.81     1.77     2.18     2.10   2.80  3.24   

   PSCA\  
0  1.22\  
1  2.59\  
2  3.59\  
3  2.95\  
4  2.78\  

[5 rows x 61 columns]

In [39]:
# remove unnecessary columns
data = data[["Date", "HomeTeam", "AwayTeam"]]

In [40]:
# map each team to a number
teams = data["HomeTeam"].unique()

In [41]:
data.shape

(380, 3)

There are 380 rows because of 380 matches in total. We optimize the first half, thus let’s take only the first 190 rows.

In [42]:
data = data.iloc[:190, :]
data.shape

(190, 3)

In [43]:
# for each date, compute which day it is in our time horizon
import datetime

format = "%d/%m/%Y"

data["Date"] = data["Date"].apply(lambda x: datetime.datetime.strptime(x, format))

# find first day in horizong
first_day = data["Date"].min()

In [45]:
# compute number of days since first day
data["Day"] = data["Date"].apply(lambda x: (x - first_day).days)

From data, get day in which each team plays (independently of whether it's home or away)

In [65]:
# create matrix of 20 rows (1 per team) and 134 columns (1 per day). Cell is 1 if team plays on that day, 0 otherwise
matrix = np.zeros((20, 134))

In [66]:
matrix.shape

(20, 134)

In [67]:
data.loc[189, "Day"]

133

In [69]:
for team in teams:
    for i, (home, away) in enumerate(zip(data["HomeTeam"], data["AwayTeam"])):
        if home == team or away == team:
            matrix[teams.tolist().index(team), data.loc[i, "Day"]] = 1

In [89]:
pd.DataFrame(matrix)

0    1    2    3    4    5    6    7    8    9    ...  124  125  126  127  \
0   1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
1   1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  1.0  0.0   
2   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
3   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
4   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
5   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
6   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
7   0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  1.0  0.0   
8   1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  1.0  0.0   
9   1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  1.0  0.0   
10  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
11  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
12  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
13  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
14  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
15  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
16  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
17  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  1.0  0.0   
18  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  1.0  0.0   
19  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   

    128  129  130  131  132  133  
0   0.0  0.0  1.0  0.0  0.0  1.0  
1   0.0  0.0  1.0  0.0  0.0  1.0  
2   0.0  0.0  1.0  0.0  0.0  1.0  
3   0.0  0.0  1.0  0.0  0.0  1.0  
4   0.0  0.0  1.0  0.0  0.0  1.0  
5   0.0  0.0  1.0  0.0  0.0  1.0  
6   0.0  0.0  1.0  0.0  0.0  1.0  
7   0.0  0.0  1.0  0.0  0.0  1.0  
8   0.0  0.0  1.0  0.0  0.0  1.0  
9   0.0  0.0  1.0  0.0  0.0  1.0  
10  0.0  0.0  1.0  0.0  0.0  1.0  
11  0.0  0.0  1.0  0.0  0.0  1.0  
12  0.0  0.0  1.0  0.0  0.0  1.0  
13  0.0  0.0  1.0  0.0  0.0  1.0  
14  0.0  0.0  1.0  0.0  0.0  1.0  
15  0.0  0.0  1.0  0.0  0.0  1.0  
16  0.0  0.0  1.0  0.0  0.0  1.0  
17  0.0  0.0  1.0  0.0  0.0  1.0  
18  0.0  0.0  1.0  0.0  0.0  1.0  
19  0.0  0.0  1.0  0.0  0.0  1.0  

[20 rows x 134 columns]

In [99]:
def GetC():

    W_ = 19
    D = 7

    C = np.zeros((D*(W_+1), D*(W_+1)))

    for t in range(W_*D+1):
        for t_ in range(W_*D+1):
            if t < t_:
                C[t, t_] = (t_ - t)**(-1)

    return C

In [100]:
GetC()

array([[0. , 1. , 0.5, ..., 0. , 0. , 0. ],
       [0. , 0. , 1. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       ...,
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ]])

In [103]:
# define total cost for each team
team_total_cost = []

# for each team 
for i in range(1, 20):
    # find index of nonzer elements, they correspond to the days in which the team plays
    team_plays = np.where(matrix[i, :] != 0)[0]

    # we will find W non-zero elements, since each team plays against all the other teams
    # for i = 1:W, return GetC()[team_plays[i], team_plays[i+1]] and sum
    team_total_cost = np.append(team_total_cost, sum([GetC()[team_plays[i], team_plays[i+1]] for i in range(1,19-1)]))


# total cost of the calendar
objective = sum(team_total_cost)

In [104]:
objective

55.75547924297924

In [108]:
for i in range(0, 20):  
    if sum(matrix[1, :]) != 19:
        print("Team " + str(i) + " plays " + str(sum(matrix[i, :])) + " times")
    else:
        print("Team " + str(i) + " all good")

Team 0 all good
Team 1 all good
Team 2 all good
Team 3 all good
Team 4 all good
Team 5 all good
Team 6 all good
Team 7 all good
Team 8 all good
Team 9 all good
Team 10 all good
Team 11 all good
Team 12 all good
Team 13 all good
Team 14 all good
Team 15 all good
Team 16 all good
Team 17 all good
Team 18 all good
Team 19 all good
